etl/analytics.py

```diff
@@ -0,0 +1,833 @@
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the chart view events.
+
+    """
+    where_clauses = []
+    if date_min:
+        where_clauses.append(f"v.day > '{date_min}'")
+    if date_max:
+        where_clauses.append(f"v.day <= '{date_max}'")
+    if chart_ids:
+        id_list = ", ".join(str(cid) for cid in chart_ids)
+        where_clauses.append(f"c.chart_id IN ({id_list})")
+
+    where_sql = "WHERE " + " AND ".join(where_clauses) if where_clauses else ""
+
+    query = f"""
+    SELECT
+        c.chart_id,
+        c.url,
+        v.day,
+        SUM(v.events) AS events
+    FROM charts c
+    JOIN grapher_views_detailed v ON c.url = v.grapher
+    {where_sql}
+    GROUP BY c.chart_id, c.url, v.day
+    ORDER BY c.chart_id, v.day ASC;
+    """
+    df = read_metabase(sql=query)
+
+    return df
+
+
+def get_chart_views_by_chart_id(
+    chart_ids: Optional[List[int]] = None,
+    date_min: str = DATE_MIN,
+    date_max: str = DATE_MAX,
+) -> pd.DataFrame:
+    """
+    Fetch number of chart views (renders) per chart from Metabase, optionally filtered by chart IDs and date range.
+
+    Parameters
+    ----------
+    chart_ids : list of int, optional
+        List of chart IDs to filter the results. If None, all charts are included.
+    date_min : str, optional
+        Minimum date to filter the results. If None, no minimum date is applied.
+    date_max : str, optional
+        Maximum date to filter the results. If None, no maximum date is applied.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the number of chart views per chart.
+
+    """
+    where_clauses = []
+    if date_min:
+        where_clauses.append(f"v.day >= '{date_min}'")
+    if date_max:
+        where_clauses.append(f"v.day <= '{date_max}'")
+    if chart_ids:
+        id_list = ", ".join(str(cid) for cid in chart_ids)
+        where_clauses.append(f"c.chart_id IN ({id_list})")
+
+    where_sql = "WHERE " + " AND ".join(where_clauses) if where_clauses else ""
+
+    query = f"""
+    SELECT
+        c.chart_id,
+        c.url,
+        c.published_at,
+        SUM(v.events) AS views
+    FROM charts c
+    JOIN grapher_views_detailed v ON c.url = v.grapher
+    {where_sql}
+    GROUP BY c.chart_id, c.url, c.published_at
+    ORDER BY views DESC
+    """
+    df_views = read_metabase(sql=query)
+
+    # To calculate the average daily views, we need to figure out the number of days for which we are counting views.
+    # This will be either the first date since we have analytics (DATE_MIN) or the publication date of the chart (if more recent than that).
+    published_at = pd.to_datetime(df_views["published_at"])
+    start_date = published_at.where(published_at > pd.to_datetime(date_min), pd.to_datetime(date_min))
+
+    # Add a column with the number of days that the views are referring to.
+    df_views["n_days"] = (pd.to_datetime(date_max) - start_date).dt.days
```

Comment:
small bug: b/c there is always a lag in analytics (of 1-2 days for page views data, longer for other external data sources like Google Search Console), the max date of the retrieved data might be less than `date_max`. Which means that the denominator in `views_daily` is larger than what it should be.
- e.g. consider a page published on 2025-05-05 with views data for 2025-05-05 to 2025-05-06 (2 days) available in `grapher_views_detailed`. If today is 2025-05-08, then the denominator for computing `views_daily` will be 4 but it should be 2.

Fix would be to retrieve the data's max date (`data_date_max = read_metabase(sql="select max(day) from grapher_views_detailed")`) and then replace `date_max` in this line with `min(date_max, data_date_max)`.

Reply:
Very good point, I've generalised this code to consider this lag.